src/NodaTime/Calendars/Wondrous/WondrousCalendarHelper.cs

```diff
@@ -0,0 +1,2204 @@
+        /// <param name="input"></param>
+        /// <returns></returns>
+        public static int WondrousMonthNum(this LocalDate input)
+        {
+            Preconditions.CheckArgument(input.Calendar == CalendarSystem.Wondrous, nameof(input), "Only valid when using the Wondrous calendar");
+
+            if (input.Month == Month18 && input.Day > DaysInMonth)
+            {
+                return 0;
+            }
+            return input.Month;
+        }
+
+        /// <summary>
+        /// Return the day of this month. 
+        /// </summary>
+        /// <remarks>Deals with days in Ayyam-i-Ha.</remarks>
+        /// <param name="input"></param>
+        /// <returns></returns>
+        public static int WondrousDayNum(this LocalDate input)
+        {
+            Preconditions.CheckArgument(input.Calendar == CalendarSystem.Wondrous, nameof(input), "Only valid when using the Wondrous calendar");
+
+            if (input.Month == Month18 && input.Day > DaysInMonth)
+            {
+                return input.Day - DaysInMonth;
+            }
+            return input.Day;
+        }
+
+
+        /// <summary>
+        /// Return the month of this date. If in Ayyam-i-Ha, returns 0.
+        /// </summary>
+        /// <remarks>Deals with Ayyam-i-Ha.</remarks>
+        /// <param name="input"></param>
+        /// <returns></returns>
+        public static int WondrousMonthNum(this LocalDateTime input)
+        {
+            return input.Date.WondrousMonthNum();
+        }
+
+        /// <summary>
+        /// Return the day of this month. 
+        /// </summary>
+        /// <remarks>Deals with days in Ayyam-i-Ha.</remarks>
+        /// <param name="input"></param>
+        /// <returns></returns>
+        public static int WondrousDayNum(this LocalDateTime input)
+        {
+            return input.Date.WondrousDayNum();
+        }
+
+        //public static WondrousYearMonthDay AsWondrousYearMonthDay(this LocalDate input)
+        //{
+        //    return new WondrousYearMonthDay(
+        //        input.Year,
+        //        input.WondrousMonthNum(),
+        //        input.WondrousDayNum());
+        //}
+
+        /// <summary>
+        /// Get a text representation of the date.
+        /// </summary>
+        /// <param name="input"></param>
+        /// <param name="format">TO BE IMPLEMENTED</param>
+        /// <returns></returns>
+        [NotNull] public static string AsWondrousString(this LocalDate input, [CanBeNull] string format = null)
+        {
+            Preconditions.CheckArgument(input.Calendar == CalendarSystem.Wondrous, nameof(input), "Only valid when using the Wondrous calendar");
+
+            // to be enhanced...
+
+            var year = input.Year;
+            var month = input.WondrousMonthNum();
+            var day = input.WondrousDayNum();
+
+            return $"{year}-{month}-{day}";
+        }
+
+        internal static WondrousYearInfo GetYearInfo(int year)
```

Comment:
This is basically 2000 lines of data. We can do a lot better than this :)
I suggest we work out a system like UmAlQuraYearMonthDayCalculator, with a base64 string containing generated data.

Am I right in saying the first value is always 4 or 5, and the second value is always 20, 21 or 22? If so, we can easily store the data for 2 years in each byte of data.

We could potentially merge like this, then I can rewrite this part, if that would be simpler than coordinating. (Is there a machine-readable form of this data somewhere so that it's easy to write a generator?)

Reply:
You got it exactly.  The source I used is [http://www.bahaidate.today/table-of-dates](http://www.bahaidate.today/table-of-dates). I've converted that to an Excel file, which I can provide a bit later. That's how I generated this code.